ArduCopter/mode_auto.cpp

```diff
@@ -1165,21 +1167,191 @@ void ModeAuto::payload_place_run()
+            break;
+        } else if (now_ms - nav_payload_place.descent_established_time_ms < descent_thrust_cal_duration_ms) {
+            // record minimum thrust for descent_thrust_cal_duration_ms
+            nav_payload_place.descent_thrust_level = MIN(nav_payload_place.descent_thrust_level, thrust_level);
+            nav_payload_place.place_start_time_ms = now_ms;
+            break;
+        } else if (thrust_level > g2.pldp_thrust_placed_fraction * nav_payload_place.descent_thrust_level) {
+            // thrust is above minimum threshold
+            nav_payload_place.place_start_time_ms = now_ms;
+            break;
+        } else if (is_positive(g2.pldp_range_finder_minimum_m)) {
+            if (!copter.rangefinder_state.enabled) {
+                // abort payload place because rangefinder is not enabled
+                nav_payload_place.state = PayloadPlaceStateType_Ascent_Start;
+                gcs().send_text(MAV_SEVERITY_WARNING, "%s PLDP_RNG_MIN set and rangefinder not enabled", prefix_str);
+                break;
+            } else if (copter.rangefinder_alt_ok() && (copter.rangefinder_state.glitch_count == 0) && (copter.rangefinder_state.alt_cm > g2.pldp_range_finder_minimum_m * 100.0)) {
+                // range finder altitude is above minimum
+                nav_payload_place.place_start_time_ms = now_ms;
+                break;
+            }
+        }
+
+        // If we get here:
+        // 1. we have reached decent velocity
+        // 2. measured the thrust level required for decent
+        // 3. detected that our thrust requirements have reduced
+        // 4. rangefinder range has dropped below minimum if set
+        // 5. place_start_time_ms has been initialised
+
+        // payload touchdown must be detected for 0.5 seconds
+
+        if (now_ms - nav_payload_place.place_start_time_ms > placed_check_duration_ms) {
+            nav_payload_place.state = PayloadPlaceStateType_Release;
+            gcs().send_text(MAV_SEVERITY_INFO, "%s payload release thrust threshold: %f", prefix_str, static_cast<double>(g2.pldp_thrust_placed_fraction * nav_payload_place.descent_thrust_level));
+        }
+        break;
+
+    case PayloadPlaceStateType_Release:
+        // Reinitialise vertical position controller to remove discontinuity due to touch down of payload
+        pos_control->init_z_controller_no_descent();
+#if AP_GRIPPER_ENABLED == ENABLED
+        if (g2.gripper.valid()) {
+            gcs().send_text(MAV_SEVERITY_INFO, "%s Releasing the gripper", prefix_str);
+            g2.gripper.release();
+            nav_payload_place.state = PayloadPlaceStateType_Releasing;
+        } else {
+            nav_payload_place.state = PayloadPlaceStateType_Delay;
+        }
+#else
+        nav_payload_place.state = PayloadPlaceStateType_Delay;
+#endif
+        break;
+
+    case PayloadPlaceStateType_Releasing:
+#if AP_GRIPPER_ENABLED == ENABLED
+        if (g2.gripper.valid() && !g2.gripper.released()) {
```

Comment:
I guess there is a tiny possibility that the pilot could move the gripper back to the unreleased position and cause the state machine to get stuck here.  I don't think it's worth protecting against because it is so unlikely.

Reply:
We decided that this would not be the best decision.